[TRTLLM-8535][feat] Support DeepSeek V3.2 with FP8 + BF16 KV cache/NVFP4 + BF16 KV cache

chang-l opened the pull request:
Co-authored with @lfr-0531 @Tracin 

This PR adds:
- DS-V3.2 support with
   - FP8 flow + BF16 KV cache
   - NVFP4 flow + BF16 KV cache
- MTP support (next_n=1)
- CUDA Graph support 
- Indexer chunked prefill for long-context request (to avoid L^2 space complexity)

TODOs:
-  Dependency PR:  https://github.com/NVIDIA/TensorRT-LLM/pull/8488

Run command: 
```
cat > extra_llm_api.yaml <<EOF
enable_attention_dp: true

moe_config:
  backend: DEEPGEMM

kv_cache_config:
  enable_block_reuse: false
EOF


trtllm-eval \
  --model ${llm_model_root}/DeepSeek-V3.2-Exp-hf/ \
  --tp_size 8 \
  --ep_size 8 \
  --extra_llm_api_options extra_llm_api.yaml \
  gsm8k

```

With indexer chunked prefill (2048)
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   | Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|------:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |96.1334|±  |0.5311|
|     |       |strict-match    |     5|exact_match|↑  |95.9818|±  |0.5409|

[10/15/2025-23:23:26] [TRT-LLM] [I] lm-eval gsm8k average accuracy: 96.06
```
W/O indexer chunking:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   | Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|------:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |95.8302|±  |0.5506|
|     |       |strict-match    |     5|exact_match|↑  |95.8302|±  |0.5506|
10/15/2025-23:31:12] [TRT-LLM] [I] lm-eval gsm8k average accuracy: 95.83
```




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added DeepSeek-V3.2 model support with sparse attention capabilities
  * Introduced DSA (Dense Sparse Attention) backend alongside existing RocketKV
  * Integrated FlashMLA kernel for optimized sparse attention operations
  * Added enable-thinking mode for reasoning models
  * Extended CLI with new options: MoE configuration, tensor parallelism, CUDA graph settings

* **Bug Fixes**
  * Fixed LayerNorm dtype casting behavior

* **Tests**
  * Added comprehensive sparse attention validation tests
  * Updated reference accuracy benchmarks for new model variants
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

lfr-0531 committed:
add a python block manager.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
add dsa config and DSATrtllmAttention, only as as wrapper for the implementation.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
add load config and load weights support for DS-v3.2.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Add indexer prefill pass and some tests

Signed-off-by: Chang Liu <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
fix DSACacheManager.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Add indexer generation kernel and indexer kv cache manager

Signed-off-by: Chang Liu <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Fix bugs for Indexer quant and cache manager and add indexer complete test

Signed-off-by: Chang Liu <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Add BF16 DSA support via FlashMLA (WIP, missing any test)

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Add FlashMLA into TRTLLM w/ basic functional test (as 3rd lib and python level api)

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Add integration test for MQA using sparse prefill kernel

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
fix DSACacheManager and add tokens_per_block to quickstart_advanced.py.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Bugfix: indexer kernel schedule metadata on host

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Slight refactor and ctx+gen batched tested

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
Some fixes and refactoring:
- Fix dtype issue
- Fix the update_resources for the indexer k cache update
- Refactor the mla_rope and load_kv_cache for both ctx and gen dsa
- Refactor the Indexer to use the kv cache manager and block offsets in metadata
- Move the metadata in Indexer to DSAtrtllmAttentionMetadata
- Add DS-v3.2 config

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Decouple indexer from DSA forward logic

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

Tracin committed:
Fix DS32 NVFP4 ckpt exclude name.

Signed-off-by: Tracin <10434017+Tracin@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Bugfix: 1-ele kscale stride(-1) issue + indexer rope + decode pass issue

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
Bugfix:
- attention module dsa forward and self.mha not defined issue
- estimate indexer k cache size issue

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
Add DSA to sparse attention example to run long sequence inputs.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
fix add_dummy_requests in DSACacheManager and pre-allocate slot_mapping_fp8 and slot_mapping_scale.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Use global topk indices

Signed-off-by: Chang Liu <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
Some fixes and refactoring:
- Pre-allocate the scheduler_metadata_buffer on device
- Move the cu_seqlen_ks and cu_seqlen_ke to metadata.prepare()
- Fix the compute_block_count in BlockManager

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
Move the torch.repeat_interleave to metadata.prepare() and fix the indexer_k_cache_block_offsets size to enable CUDA graph.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
Add more args to the eval_longbench_v1.py to support DS-v3.2.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Inter and intra-request chunking for indexer fwd

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Minor fix on indexer chunk prefill

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Acc. fix: enable hadamard transform in indexer (add hadamard pkg)

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
Add MTP support (only MTP=1)

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
fix format.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Resolve rebase conflict

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Reduce test complexity

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
rebase conflict + indexer_max_chunk_size configurable

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Add thinking mode as an option to GPQA test

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Add accuracy test into CI

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Use KVcache pool w/ zero copy to interact w/ sparse attn kernel

With formating as well

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
Bugfix: fix the python BlockManager and clean dsa.py to rm those unused host tensor.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Clean-up and use all-layer kv cache pool to the flashmla kernel

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

Superjomn reviewed (approved):
LGTM on the llmapi changes.

chang-l committed:
Address comments and cleanup

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l commented on tensorrt_llm/_torch/attention_backend/sparse/dsa.py:
```diff
@@ -0,0 +1,1433 @@
+
+        assert not kv_cache_config.enable_block_reuse, "DSA cache requires block reuse to be disabled in KV cache config"
+        self.quant_block_size = 128
+        self.index_head_dim = sparse_attn_config.index_head_dim
```
Thanks - updated.

chang-l commented:
/bot run

kaiyux commented on tensorrt_llm/_torch/configs/deepseek_v3.py:
Why do we need to introduce a copy of HF config?

yuantailing commented on tensorrt_llm/_torch/attention_backend/sparse/dsa.py:
```diff
@@ -0,0 +1,1439 @@
+        block_size = self.indexer_k_cache_manager.tokens_per_block
+        per_token_size = self.index_head_dim + self.index_head_dim // self.quant_block_size * 4
+        return self.indexer_k_cache_pool_per_layer[layer_idx].view(
+            self.num_blocks, block_size, 1, per_token_size)
```
Need to take `layer_offsets` into account. Please refer to the base KVCacheManager: https://github.com/NVIDIA/TensorRT-LLM/blob/ab4b9966b206d8faf14aadc43af2d49936a8a441/tensorrt_llm/_torch/pyexecutor/resource_manager.py#L726-L728

```suggestion
        layer_offset = self.layer_offsets[layer_idx]
        return self.indexer_k_cache_pool_per_layer[layer_offset].view(
            self.num_blocks, block_size, 1, per_token_size)
```

lfr-0531 committed:
Bugfix: fix the hang issue when running model with MTP + overlap_scheduler.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
Address comments.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 commented on tests/integration/test_lists/qa/llm_function_core.txt:
```diff
@@ -494,6 +494,8 @@ accuracy/test_llm_api_pytorch.py::TestDeepSeekR1::test_nvfp4_multi_gpus_corner_c
 accuracy/test_llm_api_pytorch.py::TestDeepSeekR1::test_fp8_blockscale[throughput]
 accuracy/test_llm_api_pytorch.py::TestDeepSeekR1::test_fp8_blockscale_chunked_prefill[latency]
 accuracy/test_llm_api_pytorch.py::TestDeepSeekR1::test_fp8_blockscale_chunked_prefill[throughput]
+accuracy/test_llm_api_pytorch.py::TestDeepSeekV32::test_baseline_fp8gemm_bf16_kv[baseline]
```
Added these two new tests to `l0_dgx_b200.yml`.

tensorrt-cicd commented:
[PR_Github #22042](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22042/) [ run ] triggered by Bot. Commit: `3e59209`

tensorrt-cicd commented:
[PR_Github #22042](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22042/) [ run ] completed with state `FAILURE`. Commit: `3e59209` 
[/LLM/main/L0_MergeRequest_PR pipeline #16620](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/16620/) completed with status: 'FAILURE'

lfr-0531 committed:
Bugfix: fix the rocket issue.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #22056](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22056/) [ run ] completed with state `SUCCESS`. Commit: `1c76df8` 
[/LLM/main/L0_MergeRequest_PR pipeline #16630](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/16630/) completed with status: 'FAILURE'

chang-l committed:
Add permissive flag for clang compiler

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #22075](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22075/) [ run ] triggered by Bot. Commit: `53a537f`

tensorrt-cicd commented:
[PR_Github #22075](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22075/) [ run ] completed with state `FAILURE`. Commit: `53a537f` 
[/LLM/main/L0_MergeRequest_PR pipeline #16646](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/16646/) completed with status: 'FAILURE'

chang-l committed:
Use gcc for flashMLA on aarch64

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l committed:
Cleanup (fix lint and add todo)

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #22087](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22087/) [ run ] triggered by Bot. Commit: `5614ea1`

chang-l committed:
Make fast-hadamard-transform optional to fix CI build failures

Problem:
CI was failing when installing requirements-dev.txt because
fast-hadamard-transform (a git package built from source) has broken
setup.py that imports wheel and torch at build-time. Pip collects
metadata for all packages before installation, causing it to run
fast-hadamard-transform's setup.py before wheel/torch are installed,
resulting in ModuleNotFoundError.

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #22098](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22098/) [ run ] triggered by Bot. Commit: `08cf10c`

tensorrt-cicd commented:
[PR_Github #22098](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22098/) [ run ] completed with state `SUCCESS`. Commit: `08cf10c` 
[/LLM/main/L0_MergeRequest_PR pipeline #16662](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/16662/) completed with status: 'FAILURE'

chang-l committed:
Skip flashMLA python binding if not there

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l commented:
/bot run

tensorrt-cicd commented:
[PR_Github #22146](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22146/) [ run ] triggered by Bot. Commit: `f39eb68`

yuxianq commented on tensorrt_llm/_torch/modules/attention.py:
```diff
@@ -1475,6 +1632,147 @@ def forward_generation(
+    def forward_sparse_mla_kvcache_bf16(
+        self,
+        q: torch.Tensor,
+        compressed_kv: torch.Tensor,
```
`compressed_kv` and `k_pe` are unused, can we remove them?

yuxianq commented on tensorrt_llm/_torch/modules/attention.py:
```diff
@@ -1475,6 +1632,147 @@ def forward_generation(
+                "flash_mla_sparse_fwd not available. Please ensure FlashMLA module is built."
+            )
+        # output: [seq, num_heads, kv_lora_rank]
+        # [seq, num_heads, kv_lora_rank]
```
These two lines seem to be repeated.

yuxianq commented on tensorrt_llm/_torch/attention_backend/sparse/dsa.py:
```diff
@@ -0,0 +1,1458 @@
+        max_batch_size: int,
+        mapping: Mapping,
+        dtype: DataType = DataType.HALF,
+        spec_config: Optional["DecodingBaseConfig"] = None,
```
`DecodingBaseConfig` is not imported, we can use `TYPE_CHECKING` trick to import it.

chang-l replied:
Thanks - updated

tensorrt-cicd commented:
[PR_Github #22194](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22194/) [ run ] triggered by Bot. Commit: `f39eb68`

chang-l commented on examples/longbench/eval_longbench_v1.py:
```diff
@@ -292,9 +317,18 @@ def initialize_llm(args: argparse.Namespace) -> Tuple[LLM, AutoTokenizer]:
         # Configure KV cache
         kv_cache_config = KvCacheConfig(
-            enable_block_reuse=False,  # RocketKV doesn't support KV cache reuse
+            # sparse attention doesn't support KV cache reuse
```
It’s a current limitation in this PR. Technically, KV-cache reuse can be supported by DSA, but the main challenge is that the additional indexer’s K-cache needs to be hashed and managed by our KV-cache manager — whereas it’s currently handled in the Python BlockManager.
See discussion here: https://github.com/NVIDIA/TensorRT-LLM/pull/8405#discussion_r2446610344

lfr-0531 commented on tensorrt_llm/_torch/models/modeling_deepseekv3.py:
```diff
@@ -383,7 +383,12 @@ def split_kv_b_proj(kv_b_proj: torch.Tensor,
+                    continue
                 else:
+                    if "indexer" in names and "mqa" in names:
+                        names.remove("mqa")
```
We can only support the indexer in trtllm attention backend now. And we also want to reuse the `sparse_attn_predict` interface in `trtllm.py` to get the `topk_indices`. Actually we're using it when supporting the sparse MLA with trtllm-gen kernels. So I prefer to leave the indexer in the backend.

yuxianq replied:
My suggestion is keeping the indexer in the attention backend and add a reference to it from MLA module, so that the module hierarchy matches with the checkpoint and we don't need this remove-mqa trick anymore. Every attention backend that supports DSA needs to have an indexer, so I think `self.indexer = self.mqa.indexer` is always correct in the DSA case.

chang-l replied:
Unfortunately, it seems even if we add `self.indexer = self.mqa.indexer` in MLA, we still can’t avoid this model-weight loading trick. 
Pytorch seems only tracks params in leaf modules — so parameter-free modules (like our reference in MLA) are skipped during weight loading (see here: https://github.com/NVIDIA/TensorRT-LLM/blob/3d74546bbd7d03b18fa8736323f242477d895319/tensorrt_llm/_torch/models/modeling_deepseekv3.py#L276

chang-l commented:
/bot run

tensorrt-cicd commented:
[PR_Github #22204](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22204/) [ run ] triggered by Bot. Commit: `f39eb68`

chang-l commented on tests/integration/defs/accuracy/test_llm_api_pytorch.py:
```diff
@@ -2315,6 +2315,73 @@ def test_fp8_blockscale_chunked_prefill(self, tp_size, pp_size, ep_size,
+            (8, 1, 8, 1, False, True, True, True, 24, "_DEFAULT"),
+        ],
+        ids=["baseline", "baseline_mtp1"])
+    def test_baseline_fp8gemm_bf16_kv(self, tp_size, pp_size, ep_size,
```
updated

chang-l commented on tensorrt_llm/_torch/models/modeling_deepseekv3.py:
```diff
@@ -1550,3 +1564,11 @@ def post_load_weights(self):
+@register_auto_model("DeepseekV32ForCausalLM")
+class DeepseekV32ForCausalLM(DeepseekV3ForCausalLM):
+
+    def load_weights(self, weights: Dict):
```
Thanks for catching it -- Fixed

chang-l committed:
Address comments + fix CI

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l commented on tests/integration/defs/disaggregated/test_configs/disagg_config_cache_reuse.yaml:
```diff
@@ -14,6 +13,7 @@ context_servers:
     enable_block_reuse: True
     enable_partial_reuse: True
     event_buffer_max_size: 1024
+    free_gpu_memory_fraction: 0.15
```
@Tabrizian Can you please take a look at this change (and confirm/approve if makes sense)? It’s unrelated to this PR — it fixes an intermittent OOM issue in context server during CI.

Tabrizian replied:
LGTM.

Tracin committed:
WAR for DSV3.2 config.

Signed-off-by: Tracin <10434017+Tracin@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

chang-l commented on tensorrt_llm/_torch/attention_backend/sparse/dsa.py:
```diff
@@ -0,0 +1,1458 @@
+        logger.warning_once(
+            "fast-hadamard-transform not available. DSA sparse attention will skip "
+            "hadamard transformation. Install with: "
+            "pip install git+https://github.com/Dao-AILab/fast-hadamard-transform.git",
```
Thanks @yuxianq! Let me file an issue to `fast_hadamard_transform` first :)

lfr-0531 committed:
add an indexer reference to the MLA attention module to simplify the weight loading and fix some rebase bugs.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 commented on tensorrt_llm/_torch/models/modeling_deepseekv3.py:
```diff
@@ -1486,6 +1491,15 @@ def __init__(self, model_config: ModelConfig[PretrainedConfig]):
+                for idx, exclude_layers in enumerate(exclude_modules):
+                    if 'indexer' in exclude_layers:
+                        exclude_modules[idx] = exclude_layers.replace(
+                            'indexer', 'mqa.indexer')
```
Added.

lfr-0531 commented on tensorrt_llm/_torch/attention_backend/sparse/dsa.py:
```diff
@@ -0,0 +1,1458 @@
+            skip_create_weights_in_init=skip_create_weights_in_init,
+            attention_chunk_size=attention_chunk_size,
+            **kwargs)
+        nn.Module.__init__(self)
```
Added the indexer into the MLA module as a reference. Keep using the `DSATrtllmAttention` here.

lfr-0531 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #22298](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22298/) [ run ] triggered by Bot. Commit: `3d74546`

yuantailing reviewed (commented):
~~It appears that `ATTRIBUTIONS-CPP-aarch64.md`, `ATTRIBUTIONS-CPP-x86_64.md`, and `ATTRIBUTIONS-Python.md` are not updated.~~
(Duplicated with the following comment, but there is no "delete" buttom)

yuantailing commented on .gitmodules:
```diff
@@ -30,3 +30,6 @@
 	branch = swapab_sm100
+[submodule "3rdparty/flash-mla"]
+	path = 3rdparty/flash-mla
+	url = https://github.com/deepseek-ai/FlashMLA.git
```
It appears that ATTRIBUTIONS-CPP-aarch64.md, ATTRIBUTIONS-CPP-x86_64.md, and ATTRIBUTIONS-Python.md are not updated.

chang-l replied:
Our attribution files actually already included flashMLA, e.g., https://github.com/NVIDIA/TensorRT-LLM/blob/f7e5aac284ca511bea676aad8746ad308b528709/ATTRIBUTIONS-CPP-aarch64.md#L5622-L5623

chang-l committed:
Address comment

Signed-off-by: Chang Liu (Enterprise Products) <9713593+chang-l@users.noreply.github.com>
Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #22415](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22415/) [ run ] triggered by Bot. Commit: `cb77f10`

tensorrt-cicd commented:
[PR_Github #22421](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22421/) [ kill ] completed with state `SUCCESS`. Commit: `6a020dd` 
Successfully killed previous jobs for commit `6a020dd`

tensorrt-cicd commented:
[PR_Github #22426](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22426/) [ run ] completed with state `SUCCESS`. Commit: `a43cf9e` 
[/LLM/main/L0_MergeRequest_PR pipeline #16901](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/16901/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #22446](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22446/) [ run ] completed with state `SUCCESS`. Commit: `a43cf9e` 
[/LLM/main/L0_MergeRequest_PR pipeline #16916](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/16916/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/16916/test-results/rerun_report.html) for details.
